### PR TITLE
ISSUE-926: Add SSL CTX cb

### PR DESCRIPTION
**Pull request by panarchus**
Add in callback configuration for SSL CTX configuration. C and C++ APIs

### CODE REVIEW

File: src/rdkafka_conf.h [hunk 1]
```diff
@@ -113,6 +113,11 @@ struct rd_kafka_conf_s {
 		char *ca_location;
 		char *crl_location;
 	} ssl;
+
```
Comment (edenhill) [inline src/rdkafka_conf.h]: put this inside the ssl struct above, for grouping.

File: src/rdkafka_transport.c [hunk 2]
```diff
@@ -762,6 +762,24 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+	 */
+	if (rk->rk_conf.ssl_ctx_cb) {
+		rd_kafka_dbg(rk, SECURITY, "SSL",
+		            "Using user-defined SSL_CTX callback");
+
+		r = rk->rk_conf.ssl_ctx_cb(rk,
+		                           (void *)ctx,
+		                           rk->rk_conf.opaque);
+
+		if (r < 0) {
```
Comment (edenhill) [inline src/rdkafka_transport.c]: the return value type is rd_kafka_resp_err_t  (use properly typed variable) and should be checked for ` != RD_KAFKA_RESP_ERR_NO_ERROR`.
Also, in this case, print the error code in the errstr and indicate where the error came from:
```
rd_snprintf(errstr, errstr_size, "User-defined SSL_CTX callback failed: %s", rd_kafka_err2str(err));
```

File: src/rdkafka.h [hunk 3]
```diff
@@ -1018,6 +1018,26 @@ void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
 
 
 /**
+ * @brief Creates/configures the SSL context using the given callback
+ *
+ * SSL contexts are usually created by providing the locations of the
+ * various files containing certificates and keys. With this set,
+ * the client has the ability to configure and set up the context
+ * to theirs needs.
```
Comment (edenhill) [inline src/rdkafka.h]: their

File: src/rdkafka.h [hunk 4]
```diff
@@ -1018,6 +1018,26 @@ void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
 
 
 /**
+ * @brief Creates/configures the SSL context using the given callback
```
Comment (edenhill) [inline src/rdkafka.h]: "Creates": the context has already been created and setup by librdkafka, so this callback can only provide additional configuration (or perhaps replace previously set config?).

I'm also wondering if maybe this should be an interceptor instead, allowing multiple sources (plugins, app) to modify the CTX.

File: src/rdkafka.h [hunk 5]
```diff
@@ -1018,6 +1018,26 @@ void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
...
+ *
+ * SSL contexts are usually created by providing the locations of the
+ * various files containing certificates and keys. With this set,
+ * the client has the ability to configure and set up the context
+ * to theirs needs.
+ *
+ * This is called once during initialization and can override any other
```
Comment (edenhill) [inline src/rdkafka.h]: "..called once per rd_kafka_t instance initialization from rd_kafka_new().."

File: src/rdkafka.h [hunk 6]
```diff
@@ -1018,6 +1018,26 @@ void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
...
+ * ssl configuration if set (called after regular rdkafka ssl
```
Comment (edenhill) [inline src/rdkafka.h]: capital SSL
spell out librdkafka

File: src/rdkafka.h [hunk 7]
```diff
@@ -1018,6 +1018,26 @@ void rd_kafka_conf_set_events(rd_kafka_conf_t *conf, int events);
...
+ * configuration).
+ *
+ */
+RD_EXPORT
+void rd_kafka_conf_set_ssl_ctx_cb (rd_kafka_conf_t *conf,
+			      rd_kafka_resp_err_t (*ssl_ctx_cb) (rd_kafka_t *rk,
+			                                         void *ssl_ctx,
```
Comment (edenhill) [inline src/rdkafka.h]: The 'void *' type here is problematic, we would really like to have this typed but that would make the header file depend on openssl.h, which we dont want.

An alternative would be to ifdef-guard this function with some define from openssl.h, allowing an application to use it if they first include openssl.h.

File: src-cpp/rdkafkacpp.h [hunk 8]
```diff
@@ -365,6 +365,21 @@ class RD_EXPORT DeliveryReportCb {
 
 
 /**
+ * @brief SSL Context Configuration callback class
+ *
+ */
+class RD_EXPORT SSLContextCb {
+ public:
+ /**
+  * @breif SSL context configuration callback
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: "breif" typo.

This class needs to be fully documented (copying the blurb from rdkafka.h).

**Comment by panarchus** [comment]
Thanks for review - I'll hopefully get to it this week (super busy atm though)

File: src/rdkafka.h (hunk 7 above)
Comment (panarchus) [inline src/rdkafka.h]: If a user compiles librdkafka and dynamically links it later, would this ifdef-guard work? The define might not be in the CU.

Comment (edenhill) [inline src/rdkafka.h]: If librdkafka compiled SSL support then the conf_set_ssl_ctx_cb symbol will be emitted,
and the application will need to be compiled using the `LIBRDKAFKA_WITH_SSL_TYPES` define to make the declaration visible.
Should be fine.

**Comment by panarchus** [comment]
Sorry it took so long to incorporate feedback, I've been entirely too busy and promptly forgot to update. I've made some changes now that I remembered: I hope this is alright!

**Pull request closed without merging**